[CI] Lifecycle-related XContentParseException in BwC tests

Actually failed on https://github.com/elastic/elasticsearch/pull/103893 but the only changes in this PR are to make the failure more likely, this can happen on `main` too.

```
» [2024-01-04T09:35:18,715][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v8.10.2-local-1] fatal error in thread [elasticsearch[v8.10.2-local-1][cluster_coordination][T#1]], exiting java.lang.AssertionError: org.elasticsearch.gateway.CorruptStateException: org.elasticsearch.xcontent.XContentParseException: [-1:23064] [index_template] failed to parse field [index_template]
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService$Writer.assertOnCommit(PersistedClusterStateService.java:1244)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService$Writer.commit(PersistedClusterStateService.java:1234)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService$Writer.writeIncrementalStateAndCommit(PersistedClusterStateService.java:976)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.GatewayMetaState$LucenePersistedState.writeClusterStateToDisk(GatewayMetaState.java:579)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.GatewayMetaState$LucenePersistedState.setLastAcceptedState(GatewayMetaState.java:562)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.coordination.CoordinationState.handlePublishRequest(CoordinationState.java:392)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.coordination.Coordinator.handlePublishRequest(Coordinator.java:469)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.coordination.PublicationTransportHandler.acceptState(PublicationTransportHandler.java:213)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.coordination.PublicationTransportHandler.handleIncomingPublishRequest(PublicationTransportHandler.java:200)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.coordination.PublicationTransportHandler.lambda$new$0(PublicationTransportHandler.java:112)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:74)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.transport.InboundHandler$1.doRun(InboundHandler.java:294)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
»  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
»  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
»  	at java.base/java.lang.Thread.run(Thread.java:1623)
»  Caused by: org.elasticsearch.gateway.CorruptStateException: org.elasticsearch.xcontent.XContentParseException: [-1:23064] [index_template] failed to parse field [index_template]
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:674)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:572)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:716)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:571)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService$Writer.assertOnCommit(PersistedClusterStateService.java:1242)
»  	... 16 more
»  Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:23064] [index_template] failed to parse field [index_template]
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:67)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2660)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:672)
»  	... 20 more
»  Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:23064] [index_template] failed to parse field [template]
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.metadata.ComposableIndexTemplate.parse(ComposableIndexTemplate.java:102)
»  	at org.elasticsearch.server@8.10.2/org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.lambda$static$1(ComposableIndexTemplateMetadata.java:50)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609)
»  	... 29 more
»  Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:23062] [template] unknown field [lifecycle]
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.lambda$errorOnUnknown$2(ObjectParser.java:105)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:299)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431)
»  	at org.elasticsearch.xcontent@8.10.2/org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609)
»  	... 36 more
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/2jr3t5m5qdlt2/tests/:qa:ccs-rolling-upgrade-remote-cluster:v8.10.2%23oldClusterTest/org.elasticsearch.upgrades.SearchStatesIT/testCanMatch


**Reproduction line:**
```
./gradlew ':qa:ccs-rolling-upgrade-remote-cluster:v8.10.2#oldClusterTest' -Dtests.class="org.elasticsearch.upgrades.SearchStatesIT" -Dtests.method="testCanMatch" -Dtests.seed=DE596670BF211DB0 -Dtests.bwc=true -Dtests.locale=es-AR -Dtests.timezone=Pacific/Ponape -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.upgrades.SearchStatesIT#testCanMatch`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testCanMatch'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.upgrades.SearchStatesIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.net.ConnectException: Connection refused

  at __randomizedtesting.SeedInfo.seed([DE596670BF211DB0:C87D533EEE92B199]:0)
  at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:934)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:304)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)
  at org.elasticsearch.test.rest.ESRestTestCase.createIndex(ESRestTestCase.java:1725)
  at org.elasticsearch.test.rest.ESRestTestCase.createIndex(ESRestTestCase.java:1678)
  at org.elasticsearch.upgrades.SearchStatesIT.testCanMatch(SearchStatesIT.java:214)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

  Caused by: java.net.ConnectException: Connection refused

    at sun.nio.ch.Net.pollConnect(Net.java:-2)
    at sun.nio.ch.Net.pollConnectNow(Net.java:682)
    at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:973)
    at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvent(DefaultConnectingIOReactor.java:174)
    at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:148)
    at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351)
    at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221)
    at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64)
    at java.lang.Thread.run(Thread.java:1583)

```

--- a/x-pack/plugin/ilm/src/main/java/module-info.java
+++ b/x-pack/plugin/ilm/src/main/java/module-info.java
@@ -18,4 +18,6 @@ module org.elasticsearch.ilm {
     provides org.elasticsearch.reservedstate.ReservedClusterStateHandlerProvider
         with
             org.elasticsearch.xpack.ilm.ReservedLifecycleStateHandlerProvider;
+
+    provides org.elasticsearch.features.FeatureSpecification with org.elasticsearch.xpack.ilm.IndexLifecycleFeatures;
 }

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycle.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycle.java
@@ -147,6 +147,7 @@ public class IndexLifecycle extends Plugin implements ActionPlugin, HealthPlugin
         ILMHistoryTemplateRegistry ilmTemplateRegistry = new ILMHistoryTemplateRegistry(
             settings,
             services.clusterService(),
+            services.featureService(),
             services.threadPool(),
             services.client(),
             services.xContentRegistry()

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleFeatures.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleFeatures.java
@@ -17,9 +17,6 @@ import java.util.Map;
 public class IndexLifecycleFeatures implements FeatureSpecification {
     @Override
     public Map<NodeFeature, Version> getHistoricalFeatures() {
-        return Map.of(
-            ILMHistoryTemplateRegistry.MANAGED_BY_DATA_STREAM_LIFECYCLE,
-            Version.V_8_12_0
-        );
+        return Map.of(ILMHistoryTemplateRegistry.MANAGED_BY_DATA_STREAM_LIFECYCLE, Version.V_8_12_0);
     }
 }

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleFeatures.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleFeatures.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ilm;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.features.FeatureSpecification;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.xpack.ilm.history.ILMHistoryTemplateRegistry;
+
+import java.util.Map;
+
+public class IndexLifecycleFeatures implements FeatureSpecification {
+    @Override
+    public Map<NodeFeature, Version> getHistoricalFeatures() {
+        return Map.of(
+            ILMHistoryTemplateRegistry.MANAGED_BY_DATA_STREAM_LIFECYCLE,
+            Version.V_8_12_0
+        );
+    }
+}

--- a/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryTemplateRegistry.java
+++ b/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryTemplateRegistry.java
@@ -8,9 +8,12 @@
 package org.elasticsearch.xpack.ilm.history;
 
 import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterChangedEvent;
 import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.features.FeatureService;
+import org.elasticsearch.features.NodeFeature;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.xcontent.NamedXContentRegistry;
 import org.elasticsearch.xpack.core.ClientHelper;
@@ -37,11 +40,13 @@ public class ILMHistoryTemplateRegistry extends IndexTemplateRegistry {
     // version 6: manage by data stream lifecycle
     // version 7: version the index template name so we can upgrade existing deployments
     public static final int INDEX_TEMPLATE_VERSION = 7;
+    public static final NodeFeature MANAGED_BY_DATA_STREAM_LIFECYCLE = new NodeFeature("ilm-history-managed-by-dsl");
 
     public static final String ILM_TEMPLATE_VERSION_VARIABLE = "xpack.ilm_history.template.version";
     public static final String ILM_TEMPLATE_NAME = "ilm-history-" + INDEX_TEMPLATE_VERSION;
 
     public static final String ILM_POLICY_NAME = "ilm-history-ilm-policy";
+    private final FeatureService featureService;
 
     @Override
     protected boolean requiresMasterNode() {
@@ -53,11 +58,13 @@ public class ILMHistoryTemplateRegistry extends IndexTemplateRegistry {
     public ILMHistoryTemplateRegistry(
         Settings nodeSettings,
         ClusterService clusterService,
+        FeatureService featureService,
         ThreadPool threadPool,
         Client client,
         NamedXContentRegistry xContentRegistry
     ) {
         super(nodeSettings, clusterService, threadPool, client, xContentRegistry);
+        this.featureService = featureService;
         this.ilmHistoryEnabled = LifecycleSettings.LIFECYCLE_HISTORY_INDEX_ENABLED_SETTING.get(nodeSettings);
     }
 
@@ -96,5 +103,10 @@ public class ILMHistoryTemplateRegistry extends IndexTemplateRegistry {
     @Override
     protected String getOrigin() {
         return ClientHelper.INDEX_LIFECYCLE_ORIGIN;
+    }
+
+    @Override
+    protected boolean isClusterReady(ClusterChangedEvent event) {
+        return featureService.clusterHasFeature(event.state(), MANAGED_BY_DATA_STREAM_LIFECYCLE);
     }
 }

--- a/x-pack/plugin/ilm/src/main/resources/META-INF/services/org.elasticsearch.features.FeatureSpecification
+++ b/x-pack/plugin/ilm/src/main/resources/META-INF/services/org.elasticsearch.features.FeatureSpecification
@@ -1,0 +1,8 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+#
+
+org.elasticsearch.xpack.ilm.IndexLifecycleFeatures

--- a/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStoreTests.java
+++ b/x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStoreTests.java
@@ -31,6 +31,7 @@ import org.elasticsearch.common.settings.ClusterSettings;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.set.Sets;
 import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.features.FeatureService;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.test.ClusterServiceUtils;
@@ -39,6 +40,7 @@ import org.elasticsearch.test.client.NoOpClient;
 import org.elasticsearch.threadpool.TestThreadPool;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xpack.ilm.IndexLifecycleFeatures;
 import org.hamcrest.Matchers;
 import org.junit.After;
 import org.junit.Before;
@@ -78,6 +80,7 @@ public class ILMHistoryStoreTests extends ESTestCase {
         ILMHistoryTemplateRegistry registry = new ILMHistoryTemplateRegistry(
             clusterService.getSettings(),
             clusterService,
+            new FeatureService(List.of(new IndexLifecycleFeatures())),
             threadPool,
             client,
             NamedXContentRegistry.EMPTY

--- a/x-pack/plugin/slm/src/main/java/module-info.java
+++ b/x-pack/plugin/slm/src/main/java/module-info.java
@@ -18,4 +18,6 @@ module org.elasticsearch.slm {
     provides org.elasticsearch.reservedstate.ReservedClusterStateHandlerProvider
         with
             org.elasticsearch.xpack.slm.ReservedLifecycleStateHandlerProvider;
+
+    provides org.elasticsearch.features.FeatureSpecification with org.elasticsearch.xpack.slm.SnapshotLifecycleFeatures;
 }

--- a/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycle.java
+++ b/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycle.java
@@ -124,6 +124,7 @@ public class SnapshotLifecycle extends Plugin implements ActionPlugin, HealthPlu
         SnapshotLifecycleTemplateRegistry templateRegistry = new SnapshotLifecycleTemplateRegistry(
             settings,
             clusterService,
+            services.featureService(),
             threadPool,
             client,
             services.xContentRegistry()

--- a/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleFeatures.java
+++ b/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleFeatures.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.slm;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.features.FeatureSpecification;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.xpack.slm.history.SnapshotLifecycleTemplateRegistry;
+
+import java.util.Map;
+
+public class SnapshotLifecycleFeatures implements FeatureSpecification {
+    @Override
+    public Map<NodeFeature, Version> getHistoricalFeatures() {
+        return Map.of(
+            SnapshotLifecycleTemplateRegistry.MANAGED_BY_DATA_STREAM_LIFECYCLE,
+            Version.V_8_12_0
+        );
+    }
+}

--- a/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleFeatures.java
+++ b/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleFeatures.java
@@ -17,9 +17,6 @@ import java.util.Map;
 public class SnapshotLifecycleFeatures implements FeatureSpecification {
     @Override
     public Map<NodeFeature, Version> getHistoricalFeatures() {
-        return Map.of(
-            SnapshotLifecycleTemplateRegistry.MANAGED_BY_DATA_STREAM_LIFECYCLE,
-            Version.V_8_12_0
-        );
+        return Map.of(SnapshotLifecycleTemplateRegistry.MANAGED_BY_DATA_STREAM_LIFECYCLE, Version.V_8_12_0);
     }
 }

--- a/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotLifecycleTemplateRegistry.java
+++ b/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotLifecycleTemplateRegistry.java
@@ -8,10 +8,13 @@
 package org.elasticsearch.xpack.slm.history;
 
 import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterChangedEvent;
 import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.features.FeatureService;
+import org.elasticsearch.features.NodeFeature;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.xcontent.NamedXContentRegistry;
 import org.elasticsearch.xpack.core.ilm.IndexLifecycleMetadata;
@@ -44,11 +47,13 @@ public class SnapshotLifecycleTemplateRegistry extends IndexTemplateRegistry {
     // version 6: manage by data stream lifecycle
     // version 7: version the index template name so we can upgrade existing deployments
     public static final int INDEX_TEMPLATE_VERSION = 7;
+    public static final NodeFeature MANAGED_BY_DATA_STREAM_LIFECYCLE = new NodeFeature("slm-history-managed-by-dsl");
 
     public static final String SLM_TEMPLATE_VERSION_VARIABLE = "xpack.slm.template.version";
     public static final String SLM_TEMPLATE_NAME = ".slm-history-" + INDEX_TEMPLATE_VERSION;
 
     public static final String SLM_POLICY_NAME = "slm-history-ilm-policy";
+    private final FeatureService featureService;
 
     @Override
     protected boolean requiresMasterNode() {
@@ -60,11 +65,13 @@ public class SnapshotLifecycleTemplateRegistry extends IndexTemplateRegistry {
     public SnapshotLifecycleTemplateRegistry(
         Settings nodeSettings,
         ClusterService clusterService,
+        FeatureService featureService,
         ThreadPool threadPool,
         Client client,
         NamedXContentRegistry xContentRegistry
     ) {
         super(nodeSettings, clusterService, threadPool, client, xContentRegistry);
+        this.featureService = featureService;
         slmHistoryEnabled = SLM_HISTORY_INDEX_ENABLED_SETTING.get(nodeSettings);
     }
 
@@ -114,5 +121,10 @@ public class SnapshotLifecycleTemplateRegistry extends IndexTemplateRegistry {
 
         boolean allPoliciesPresent = maybePolicies.map(policies -> policies.keySet().containsAll(policyNames)).orElse(false);
         return allTemplatesPresent && allPoliciesPresent;
+    }
+
+    @Override
+    protected boolean isClusterReady(ClusterChangedEvent event) {
+        return featureService.clusterHasFeature(event.state(), MANAGED_BY_DATA_STREAM_LIFECYCLE);
     }
 }

--- a/x-pack/plugin/slm/src/main/resources/META-INF/services/org.elasticsearch.features.FeatureSpecification
+++ b/x-pack/plugin/slm/src/main/resources/META-INF/services/org.elasticsearch.features.FeatureSpecification
@@ -1,0 +1,8 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+#
+
+org.elasticsearch.xpack.slm.SnapshotLifecycleFeatures

--- a/x-pack/plugin/slm/src/test/java/org/elasticsearch/xpack/slm/history/SnapshotLifecycleTemplateRegistryTests.java
+++ b/x-pack/plugin/slm/src/test/java/org/elasticsearch/xpack/slm/history/SnapshotLifecycleTemplateRegistryTests.java
@@ -26,6 +26,7 @@ import org.elasticsearch.cluster.node.DiscoveryNodes;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.TriFunction;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.features.FeatureService;
 import org.elasticsearch.test.ClusterServiceUtils;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.client.NoOpClient;
@@ -46,6 +47,7 @@ import org.elasticsearch.xpack.core.ilm.OperationMode;
 import org.elasticsearch.xpack.core.ilm.RolloverAction;
 import org.elasticsearch.xpack.core.ilm.TimeseriesLifecycleType;
 import org.elasticsearch.xpack.core.ilm.action.PutLifecycleAction;
+import org.elasticsearch.xpack.slm.SnapshotLifecycleFeatures;
 import org.junit.After;
 import org.junit.Before;
 
@@ -98,7 +100,14 @@ public class SnapshotLifecycleTemplateRegistryTests extends ESTestCase {
             )
         );
         xContentRegistry = new NamedXContentRegistry(entries);
-        registry = new SnapshotLifecycleTemplateRegistry(Settings.EMPTY, clusterService, threadPool, client, xContentRegistry);
+        registry = new SnapshotLifecycleTemplateRegistry(
+            Settings.EMPTY,
+            clusterService,
+            new FeatureService(List.of(new SnapshotLifecycleFeatures())),
+            threadPool,
+            client,
+            xContentRegistry
+        );
     }
 
     @After
@@ -113,6 +122,7 @@ public class SnapshotLifecycleTemplateRegistryTests extends ESTestCase {
         SnapshotLifecycleTemplateRegistry disabledRegistry = new SnapshotLifecycleTemplateRegistry(
             settings,
             clusterService,
+            new FeatureService(List.of(new SnapshotLifecycleFeatures())),
             threadPool,
             client,
             xContentRegistry

